### ISSUE TITLE
facegeometry , facemesh : how to get real world 3D coordinates for face landmarks

### ISSUE THREAD

**Issue by razor123max** [google/mediapipe#3362]
<em>Please make sure that this is a feature request.</em>

**System information** (Please provide as much relevant information as possible)
arm jestson build

- MediaPipe Solution (you are using): face mesh
- Programming language : C++

Need help in getting real-world 3d coordinates for the facemesh landmarks.

How do I access the realworld 3d cordinates using face geometry module? as per my understanding 

"FaceGeometryPipelineCalculator | Extracts face 3D transform for multiple faces from a vector of landmark lists."
while doing this in 1st step it converts camera landmarks to real world 3 D landmark and use it to calculate pose transformation matrix, but my question is is there any way we can get access to realworld 3d landmark list ?
or any other way to get landmark coordinates?

[FaceGeometryPipelineCalculator](https://github.com/google/mediapipe/tree/master/mediapipe/modules/face_geometry/geometry_pipeline_calculator.cc)	Extracts face 3D transform for multiple faces from a vector of landmark lists.

Thank you,

**Comment by razor123max** [comment]
hi @sureshdagooglecom ,
I think this approach should give me position of the each landmark in image frame (approximate pixel number)?

I am specifically looking for the 3D world coordinates for each of the landmarks (in centimetres),

The documentation states in face transform module specifically in the component section, 1st bullet point 
"Face landmark screen coordinates are converted into the Metric 3D space coordinates", as i understand expect use of "transform pipeline" is to get  A vector of face geometry data. 
But is there any way to use this data to get back landmarks in real-world 3D coordinates in centimetre ? as in 1st step to calculate this pose matrix internally in 1st step landmarks are converted to metric 3D space?

I am trying to use c++ version. Thanks.

https://github.com/google/mediapipe/tree/master/mediapipe/modules/face_geometry

**Comment by andrechen** [comment]
@razor123max [this](https://github.com/google/mediapipe/blob/master/mediapipe/modules/face_geometry/libs/geometry_pipeline.cc) is what you're looking for. I did some simplification and optimization of the algorithm, maybe i'll post it on my [fork](https://github.com/andrechen/mediapipe) when possible. 

Did anyone notice the 'face_effect' sample (the best demo to illustrate this)  has gone?

https://www.linkedin.com/posts/andre-chen-4a8a7592_another-much-better-3d-face-reconstruction-activity-6918921512129355776-NE2I?utm_source=linkedin_share&utm_medium=member_desktop_web

**Comment by razor123max** [comment]
@andrechen thanks, could you pleae update your simplifications on your fork. 

currently mediapipe/mediapipe/modules/face_geometry/libs/geometry_pipeline.h does not have any thing that i can use to get back 3D world coordinates. I can see they are being calculated in modules/face_geometry/libs/geometry_pipeline.cc but are not available outside?

Let me know how did you simplified this?

or any approach to recalculate them based on face pose matrix.

**Comment by razor123max** [comment]
So, As I understand if i simply **multiply pose_transform_matrix**  with each of the face normalized landmarks, i should get them in 3D real world frame?

**Comment by andrechen** [comment]
Not really.
The pose_transform_matrix approximately transforms the [canonical face model](https://github.com/google/mediapipe/blob/master/mediapipe/modules/face_geometry/data/canonical_face_model.obj) to the face mesh deduced from normalized landmarks. Meaning, if you have a 3D model well-aligned with canonical face model in the same coordinate system (like a pair of glasses), you use pose_transform_matrix to transform the model so the model after transformed will fit well on constructed face mesh.  On Windows, you can double-click **canonical_face_model.obj** to feel it in 3D. The deduced face mesh is the other member of FaceGeometry structure.

**Comment by razor123max** [comment]
Okay, So currently I get the FaceGeometry from "FaceGeometryFromLandmarks" calculator, how do i access the deduced face mesh from FaceGeometry? with this facemesh can i get landmarks in cm?

**Comment by andrechen** [comment]
2340 floats = 468 vertices. Each vertex contains 5 floats.  The first 3 floats should be XYZ coordinates in centimeter, still 2 floats are UV, or normalized texture coordinate from canonical face model.

**Comment by razor123max** [comment]
ahh! that's why i was getting almost constant values for landmarks ... anyway i had tries multiplying them with pose transform matrix, to get back the values in 3D space. I am using face-mesh with iris landmarks as well, is there anyway to get iris normalized landmarks in cm?

**Comment by andrechen** [comment]
geometry_pipeline.cc [L127](https://github.com/google/mediapipe/blob/4a20e9909d55838d5630366ce719844cf06ae85c/mediapipe/modules/face_geometry/libs/geometry_pipeline.cc#L127), with_attention model, 478  landmarks will fail this **RET_CHECK_EQ(478, 468)**. Surely, change **RET_CHECK_EQ**() to **RET_CHECK_GE**() will give it a pass, but ScreenToMetricSpaceConverter::Convert() still processes 468 landmarks only. 
To get these 10 extra metric landmarks, maybe you can take few 3D landmarks around the eye, and use iris' 2D normalized landmark coordinates to do interpolations.

**Comment by razor123max** [comment]
Thanks for the suggestion, i tried with simple linear scaling between landmark 386 and center of iris, it has some scaling issues obviously z axis does not work. If you can try to give me small snippet using single landmark that will help.

**Comment by andrechen** [comment]
just pushed [face_effect](https://github.com/andrechen/mediapipe/tree/master/visualstudio/face_effect) demo. Not thoroughly checked, but it does work! (finally works on windows + visual studio). Knowing the estimated metric 3D landmarks may depend on camera FOV settings, your face shape/scale difference compare to canonical face model, this demo uses some pre-defined key inputs to change camera FOV settings, different model scaling plus different models (with different primitive types) to show.

I also added a [face_geometry test calculator](https://github.com/andrechen/mediapipe/blob/a2d29d53c01c6201e3c5f221abfe6aabd7f1ec3e/visualstudio/face_mesh/face_geometry_module_test.cc#L166) to face_mesh demo, you can debug it with visual studio to check if **face_geometry** module really works for you.

**Comment by razor123max** [comment]
Thanks @andrechen, your code does help in understanding the module facegeometry, it is inline with what we have discussing over this thread, for now i was not able to get the iris landmarks in 3d space by using some nearby eye border landmark, the simple scaling factor based on normalized eye border landmark result in some arbitrary landmarks, i thing face landmark and iris landmarks are normalized with respect to different base. If you find something in this line please let me know...

**Comment by andrechen** [comment]
The problems of face_geometry modules are 1) a huge waste on computing zero-weighted errors (total 435 of 468 has zero weights). 2) the converted landmarks must be 1-1 map to canonical face model, so only 468s will be converted. Actually the rest 10 iris landmarks can use the same rules to convert to metric space. 3) most parameters during the computation are always the same, meaning, they should be pre-computed one time, not the real-time. 

In short, the easiest way to achieve your goal is to modify the face_geometry module. It's not that hard, just like [this](https://github.com/andrechen/mediapipe/blob/ca7ddc37c77fa237576161b054c9f795d794a77f/visualstudio/face_effect/face_effect_renderer.cc#L110).

**Comment by dandingol03** [comment]
> just pushed [face_effect](https://github.com/andrechen/mediapipe/tree/master/visualstudio/face_effect) demo. Not thoroughly checked, but it does work! (finally works on windows + visual studio). Knowing the estimated metric 3D landmarks may depend on camera FOV settings, your face shape/scale difference compare to canonical face model, this demo uses some pre-defined key inputs to change camera FOV settings, different model scaling plus different models (with different primitive types) to show.
> 
> I also added a [face_geometry test calculator](https://github.com/andrechen/mediapipe/blob/a2d29d53c01c6201e3c5f221abfe6aabd7f1ec3e/visualstudio/face_mesh/face_geometry_module_test.cc#L166) to face_mesh demo, you can debug it with visual studio to check if **face_geometry** module really works for you.

Why `gives us the idea that after the first step, Zave is should be 1.0.`. This comment appear in [face_effect](https://github.com/andrechen/mediapipe/tree/master/visualstudio/face_effect) demo
